예외 테스트

1. @Test(expected)

예외 타입만 확인 가능

```java
@Test(expected = UsernameNotFoundException.class)
    public void findByUsername_fail(){
        accountService.loadUserByUsername("Non Existing Username");
    }
```

2. try-catch

예외 타입과 메시지 확인 가능.

하지만 코드가 다소 복잡.

```java
@Test
    public void findByUsername_fail(){
        String username = "NonExistingUsername";
        try {
            accountService.loadUserByUsername(username);
            fail("Non Error");
        } catch (UsernameNotFoundException e){
            assertThat(e.getMessage()).containsSequence(username);
        }
    }
```


3. @Rule ExpectedException

코드는 간결하면서 예외 타입과 메시지 모두 확인 가능

```java
@Rule
    public ExpectedException expectedException = ExpectedException.none();
    
@Test
    public void findByUsername_fail(){
        String username = "NonExistingUsername";

        expectedException.expect(UsernameNotFoundException.class);
        expectedException.expectMessage(Matchers.containsString(username));

        accountService.loadUserByUsername(username);
    }
```

주의) 

- expectedException 이 테스트 대상 실행 전에 명시가 되어야 한다.

-  expectedException 이 나중에 나올 경우엔, 테스트 대상이 실행되면서 에러가 발생하여 fail 되어 끝나버린다.

---

따라서, 테스트 결과는 성공이거나 예측 결과와 다른 경우, 두가지여야 잘 작성 된 것이다.

테스트 결과가 아예 실패라는 것은, 발생한 에러를 처리하지 못했다는 것이다. 이는 잘못 작성된 테스트인 것이다.
